contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol

```diff
@@ -0,0 +1,710 @@
+     */
+    function initialize(address[] calldata _rewardTokenAddresses)
+        external
+        onlyGovernor
+        initializer
+    {
+        address[] memory pTokens = new address[](1);
+        pTokens[0] = pool;
+
+        address[] memory _assets = new address[](1);
+        _assets[0] = ws;
+
+        InitializableAbstractStrategy._initialize(
+            _rewardTokenAddresses,
+            _assets,
+            pTokens
+        );
+
+        _approveBase();
+    }
+
+    /***************************************
+                    Deposit
+    ****************************************/
+
+    /**
+     * @notice Deposit Wrapped S (wS) into the SwapX pool
+     * @param _asset Address of Wrapped S (wS) contract.
+     * @param _amount Amount of Wrapped S (wS) to deposit.
+     */
+    function deposit(address _asset, uint256 _amount)
+        external
+        override
+        onlyVault
+        nonReentrant
+        skimPool
+    {
+        _deposit(_asset, _amount);
+    }
+
+    function _deposit(address _wS, uint256 _wsAmount) internal {
+        require(_wsAmount > 0, "Must deposit something");
+        require(_wS == ws, "Unsupported asset");
+
+        // Calculate the required amount of OS to mint based on the wS amount.
+        uint256 osDepositAmount = _calcTokensToMint(_wsAmount);
+
+        // Mint the required OS tokens to this strategy
+        IVault(vaultAddress).mintForStrategy(osDepositAmount);
+
+        // Add wS and OS liquidity to the pool and stake in gauge
+        _depositToPoolAndGauge(_wsAmount, osDepositAmount);
+
+        // Ensure solvency of the vault
+        _solvencyAssert();
+
+        // Emit event for the deposited wS tokens
+        emit Deposit(_wS, pool, _wsAmount);
+        // Emit event for the minted OS tokens
+        emit Deposit(os, pool, osDepositAmount);
+    }
+
+    /**
+     * @notice Deposit the strategy's entire balance of Wrapped S (wS) into the pool
+     */
+    function depositAll() external override onlyVault nonReentrant skimPool {
+        uint256 balance = IERC20(ws).balanceOf(address(this));
+        if (balance > 0) {
+            _deposit(ws, balance);
+        }
+    }
+
+    /***************************************
+                    Withdraw
+    ****************************************/
+
+    /**
+     * @notice Withdraw wS and OS from the SwapX pool, burn the OS,
+     * and transfer the wS to the recipient.
+     * @param _recipient Address to receive withdrawn asset which is normally the Vault.
+     * @param _asset Address of the Wrapped S (wS) contract.
+     * @param _wsAmount Amount of Wrapped S (wS) to withdraw.
+     */
+    function withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _wsAmount
+    ) external override onlyVault nonReentrant skimPool {
+        require(_wsAmount > 0, "Must withdraw something");
+        require(_asset == ws, "Unsupported asset");
+
```

Comment:
We should add a require here: 
```
require(_recipient == vaultAddress, "Only withdraw to vault allowed");
```

Reply:
VaultCore's `redeem` code does allow withdrawing directly to the redeemer. But that's not relevant for Sonic as `redeem` has been disabled.

```Solidity
// Nothing in Vault, but something in Strategy, send from there
IStrategy strategy = IStrategy(strategyAddr);
strategy.withdraw(msg.sender, assetAddr, outputs[i]);
```

I'm inclined to keep the behaviour between the strategies consistent and allow the Vault to withdraw to a recipient.
I can be convinced the other way, though.

Reply:
I guess it comes down to, can/will this ever be an asset default strategy. With AerodromAMO we knew that it won't be, so we added [the require check](https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol#L889). 

I think any strategy operating on a pool will probably never be an asset default one, and for that reason we could add a require here. Again since only vault calls this code it is a defensive programming measure that is of low importance. The require would basically only prevent the governance mistakingly setting the SwapxAMO as a default strategy. I will leave the decision up to you.

Reply:
I'm happy to do the same as the Aerodrome AMO. I like being consistent :-)

Given deposits and withdrawals can be front-run, we definitely don't want the AMO strategy set as the default strategy in the vault.

I'll add the require